trunk/cl_dzip.c

```diff
@@ -130,33 +141,45 @@ DZip_StartExtract (dzip_context_t *ctx, const char *name, FILE **demo_file_p)
 		return DZIP_ALREADY_EXTRACTING;
 	}
 
-	// Set `name` to the base name of the dzip, and `dem_basedir` to the rest of the path,
-	// and `dz_name` to the concatenation of the two.
+	// Set `base_name` to the base name of the dzip, and `dem_basedir` to the
+	// rest of the path, and `dz_name` to the concatenation of the two.
 	if (strstr(name, "..") == name)
 	{
-		Q_snprintfz (dem_basedir, sizeof(dem_basedir), "%s%s", com_basedir, name + 2);
-		p = strrchr (dem_basedir, '/');
-		*p = 0;
-		name = strrchr (name, '/') + 1;	// we have to cut off the path for the name
+		Q_snprintfz (dz_name, sizeof(dz_name), "%s%s", com_basedir, name + 2);
 	}
 	else
 	{
-		Q_strncpyz (dem_basedir, com_gamedir, sizeof(dem_basedir));
```

Comment:
There was a bug in this case relating to subdirs.  For example, if you have a dzip at `<base dir>/joequake/er/e1m1_029.dz`  and you attempted to play it with `playdemo er/e1m1_029.dz`, then the logic would set `dem_basedir` and `name` as `<base dir>/joequake` and `er/e1m1_029.dz` respectively.   This means the dzip would be extracted into the `joequake` directory, but would attempt to play  `joequake/er/e1m1_029.dem`.

I've changed the logic to make `base_name` just the file name with no directories, and `dem_basedir` to be  all the directories.  `name` is kept as the path provided by the user.